File: CloudSeed/DelayLine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AudioLib;
using AudioLib.Modules;
using AudioLib.TF;

namespace CloudSeed
{
	public class DelayLine
	{
		private readonly ModulatedDelay delay;
		private readonly AllpassDiffuser diffuser;
		private readonly Biquad lowShelf;
		private readonly Biquad highShelf;
		private readonly Lp1 lowPass;
		private readonly double[] tempBuffer;
		private readonly double[] filterOutputBuffer;

		private double feedback;
		private int samplerate;

		public bool DiffuserEnabled;
		public bool LowShelfEnabled;
		public bool HighShelfEnabled;
		public bool CutoffEnabled;

		public DelayLine(int bufferSize, int samplerate)
		{
			delay = new ModulatedDelay(bufferSize, 10000);
			diffuser = new AllpassDiffuser(bufferSize, samplerate) { ModulationEnabled = false };
			tempBuffer = new double[bufferSize];
			filterOutputBuffer = new double[bufferSize];

			lowPass = new Lp1(samplerate);
			lowShelf = new Biquad(Biquad.FilterType.LowShelf, samplerate) { Slope = 1.0, GainDB = -20, Frequency = 20 };
			highShelf = new Biquad(Biquad.FilterType.HighShelf, samplerate) { Slope = 1.0, GainDB = -20, Frequency = 19000 };
			lowPass.CutoffHz = 1000;
			lowShelf.Update();
			highShelf.Update();
			Samplerate = samplerate;
		}

		public int Samplerate
		{
			get { return samplerate; }
			set
			{
				samplerate = value;
				diffuser.Samplerate = samplerate;
				lowPass.Samplerate = samplerate;
				lowShelf.Samplerate = samplerate;
				highShelf.Samplerate = samplerate;
			}
		}

		public double[] DiffuserSeeds
		{
			get { return diffuser.Seeds; }
			set { diffuser.Seeds = value; }
		}

		public void SetDelay(int delaySamples)
		{
			delay.SampleDelay = delaySamples;
		}

		public void SetFeedback(double feedb)
		{
			feedback = feedb;
		}
		
		public void SetDiffuserDelay(int delaySamples)
		{
			diffuser.SetDelay(delaySamples);
		}

		public void SetDiffuserFeedback(double feedb)
		{
			diffuser.SetFeedback(feedb);
		}

		public void SetDiffuserStages(int stages)
		{
			diffuser.Stages = stages;
		}

		public void SetLowShelfGain(double gain)
		{
			lowShelf.Gain = gain;
			lowShelf.Update();
		}

		public void SetLowShelfFrequency(double frequency)
		{
			lowShelf.Frequency = frequency;
			lowShelf.Update();
		}

		public void SetHighShelfGain(double gain)
		{
			highShelf.Gain = gain;
			highShelf.Update();
		}
		
		public void SetHighShelfFrequency(double frequency)
		{
			highShelf.Frequency = frequency;
			highShelf.Update();
		}

		public void SetCutoffFrequency(double frequency)
		{
			lowPass.CutoffHz = frequency;
		}

		public void SetModAmount(double amount)
		{
			delay.ModAmount = amount;
		}

		public void SetModRate(double rate)
		{
			delay.ModRate = rate;
		}

		public double[] Output { get { return delay.Output; } }

		public void Process(double[] input, int sampleCount)
		{
			var feedbackBuffer = DiffuserEnabled ? diffuser.Output : filterOutputBuffer;
			
			for (int i = 0; i < sampleCount; i++)
				tempBuffer[i] = input[i] + feedbackBuffer[i] * feedback;

			delay.Process(tempBuffer, sampleCount);
			delay.Output.Copy(tempBuffer, sampleCount);

			if (LowShelfEnabled)
				lowShelf.Process(tempBuffer, tempBuffer, sampleCount);
			if (HighShelfEnabled)
				highShelf.Process(tempBuffer, tempBuffer, sampleCount);
			if (CutoffEnabled)
				lowPass.Process(tempBuffer, tempBuffer, sampleCount);

			tempBuffer.Copy(filterOutputBuffer, sampleCount);

			if (DiffuserEnabled)
			{
				diffuser.Process(filterOutputBuffer, sampleCount);
			}
		}

		public void ClearBuffers()
		{
			delay.ClearBuffers();
			diffuser.ClearBuffers();
			lowShelf.ClearBuffers();
			highShelf.ClearBuffers();
			lowPass.Output = 0;
			tempBuffer.Zero();
			filterOutputBuffer.Zero();
		}
	}
}

File: CloudSeed.Native/DelayLine.cpp
#include "DelayLine.h"
#include "Utils.h"
#include <iostream>

namespace CloudSeed
{
	DelayLine::DelayLine(int bufferSize, int samplerate)
		: lowPass(samplerate)
		, delay(bufferSize, 10000)
		, diffuser(bufferSize, samplerate)
		, lowShelf(AudioLib::Biquad::FilterType::LowShelf, samplerate)
		, highShelf(AudioLib::Biquad::FilterType::HighShelf, samplerate)
	{
		this->SampleResolution = 32;
		this->Undersampling = 1;

		this->bufferSize = bufferSize;
		tempBuffer = new double[bufferSize];
		mixedBuffer = new double[bufferSize];
		filterOutputBuffer = new double[bufferSize];
		
		lowShelf.Slope = 1.0;
		lowShelf.SetGainDb(-20);
		lowShelf.Frequency = 20;

		highShelf.Slope = 1.0;
		highShelf.SetGainDb(-20);
		highShelf.Frequency = 19000;

		lowPass.SetCutoffHz(1000);
		lowShelf.Update();
		highShelf.Update();
		SetSamplerate(samplerate);
	}

	DelayLine::~DelayLine()
	{
		std::cout << "Deleting DelayLine " << std::endl;
		delete tempBuffer;
		delete mixedBuffer;
		delete filterOutputBuffer;
	}

	int DelayLine::GetSamplerate()
	{
		return samplerate;
	}

	void DelayLine::SetSamplerate(int samplerate)
	{
		this->samplerate = samplerate;
		diffuser.SetSamplerate(samplerate);
		lowPass.SetSamplerate(samplerate);
		lowShelf.SetSamplerate(samplerate);
		highShelf.SetSamplerate(samplerate);
	}

	vector<double> DelayLine::GetDiffuserSeeds()
	{
		return diffuser.GetSeeds();
	}

	void DelayLine::SetDiffuserSeeds(vector<double> seeds)
	{
		diffuser.SetSeeds(seeds);
	}

	void DelayLine::SetDelay(int delaySamples)
	{
		delay.SampleDelay = delaySamples;
	}

	void DelayLine::SetFeedback(double feedb)
	{
		feedback = feedb;
	}

	void DelayLine::SetDiffuserDelay(int delaySamples)
	{
		diffuser.SetDelay(delaySamples);
	}

	void DelayLine::SetDiffuserFeedback(double feedb)
	{
		diffuser.SetFeedback(feedb);
	}

	void DelayLine::SetDiffuserStages(int stages)
	{
		diffuser.Stages = stages;
	}

	void DelayLine::SetLowShelfGain(double gain)
	{
		lowShelf.SetGain(gain);
		lowShelf.Update();
	}

	void DelayLine::SetLowShelfFrequency(double frequency)
	{
		lowShelf.Frequency = frequency;
		lowShelf.Update();
	}

	void DelayLine::SetHighShelfGain(double gain)
	{
		highShelf.SetGain(gain);
		highShelf.Update();
	}

	void DelayLine::SetHighShelfFrequency(double frequency)
	{
		highShelf.Frequency = frequency;
		highShelf.Update();
	}

	void DelayLine::SetCutoffFrequency(double frequency)
	{
		lowPass.SetCutoffHz(frequency);
	}

	void DelayLine::SetLineModAmount(double amount)
	{
		delay.ModAmount = amount;
	}

	void DelayLine::SetLineModRate(double rate)
	{
		delay.ModRate = rate;
	}

	void DelayLine::SetDiffuserModAmount(double amount)
	{
		diffuser.SetModulationEnabled(amount > 0.0);
		diffuser.SetModAmount(amount);
	}

	void DelayLine::SetDiffuserModRate(double rate)
	{
		diffuser.SetModRate(rate);
	}

	void DelayLine::SetInterpolationEnabled(bool value)
	{
		diffuser.SetInterpolationEnabled(value);
	}

	double* DelayLine::GetOutput()
	{
		if (LateStageTap)
		{
			if (DiffuserEnabled)
				return diffuser.GetOutput();
			else
				return mixedBuffer;
		}
		else
		{
			return delay.GetOutput();
		}
	}

	void DelayLine::Process(double* input, int sampleCount)
	{
		for (int i = 0; i < sampleCount; i++)
			mixedBuffer[i] = input[i] + filterOutputBuffer[i] * feedback;

		if (LateStageTap)
		{
			if (DiffuserEnabled)
			{
				diffuser.Process(mixedBuffer, sampleCount);
				Utils::BitcrushAndReduce(diffuser.GetOutput(), mixedBuffer, sampleCount, Undersampling, SampleResolution);
				delay.Process(mixedBuffer, sampleCount);
			}
			else
			{
				Utils::BitcrushAndReduce(mixedBuffer, mixedBuffer, sampleCount, Undersampling, SampleResolution);
				delay.Process(mixedBuffer, sampleCount);
			}

			Utils::Copy(delay.GetOutput(), tempBuffer, sampleCount);
		}
		else
		{
			
			if (DiffuserEnabled)
			{
				Utils::BitcrushAndReduce(mixedBuffer, mixedBuffer, sampleCount, Undersampling, SampleResolution);
				delay.Process(mixedBuffer, sampleCount);
				diffuser.Process(delay.GetOutput(), sampleCount);
				Utils::Copy(diffuser.GetOutput(), tempBuffer, sampleCount);
			}
			else
			{
				Utils::BitcrushAndReduce(mixedBuffer, mixedBuffer, sampleCount, Undersampling, SampleResolution);
				delay.Process(mixedBuffer, sampleCount);
				Utils::Copy(delay.GetOutput(), tempBuffer, sampleCount);
			}
		}

		if (LowShelfEnabled)
			lowShelf.Process(tempBuffer, tempBuffer, sampleCount);
		if (HighShelfEnabled)
			highShelf.Process(tempBuffer, tempBuffer, sampleCount);
		if (CutoffEnabled)
			lowPass.Process(tempBuffer, tempBuffer, sampleCount);

		Utils::Copy(tempBuffer, filterOutputBuffer, sampleCount);
	}

	void DelayLine::ClearDiffuserBuffer()
	{
		diffuser.ClearBuffers();
	}

	void DelayLine::ClearBuffers()
	{
		delay.ClearBuffers();
		diffuser.ClearBuffers();
		lowShelf.ClearBuffers();
		highShelf.ClearBuffers();
		lowPass.Output = 0;

		for (int i = 0; i < bufferSize; i++)
		{
			tempBuffer[i] = 0.0;
			filterOutputBuffer[i] = 0.0;
		}
	}
}
